types/src/transaction/mod.rs

```diff
@@ -529,15 +674,153 @@ pub enum TransactionPayload {
     /// transaction as the multisig account.
     Multisig(Multisig),
+    /// A new transaction payload format with support for versioning.
+    /// Contains an executable (script/entry function) along with extra configuration.
+    /// Once this new format is fully rolled out, above payload variants will be deprecated.
+    Payload(TransactionPayloadInner),
+}
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq, Serialize, Deserialize)]
+pub enum TransactionPayloadInner {
+    V1 {
+        executable: TransactionExecutable,
+        extra_config: TransactionExtraConfig,
+    },
+}
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq, Serialize, Deserialize)]
+pub enum TransactionExecutable {
+    Script(Script),
+    EntryFunction(EntryFunction),
+    Empty,
+}
+
+impl TransactionExecutable {
+    pub fn is_empty(&self) -> bool {
+        matches!(self, Self::Empty)
+    }
+
+    pub fn is_script(&self) -> bool {
+        matches!(self, Self::Script(_))
+    }
+
+    pub fn is_entry_function(&self) -> bool {
+        matches!(self, Self::EntryFunction(_))
+    }
+}
+
+#[derive(Clone, Debug, Hash, Eq, PartialEq, Serialize, Deserialize)]
+pub enum TransactionExtraConfig {
+    V1 {
+        multisig_address: Option<AccountAddress>,
+        // None for regular transactions
+        // Some(nonce) for orderless transactions
+        replay_protection_nonce: Option<u64>,
+    },
 }
 
-impl TransactionPayload {
+impl TransactionPayloadWrapper {
     pub fn into_entry_function(self) -> EntryFunction {
         match self {
             Self::EntryFunction(f) => f,
             payload => panic!("Expected EntryFunction(_) payload, found: {:#?}", payload),
         }
     }
+
+    pub fn has_nonce(&self) -> bool {
+        match self {
+            Self::Payload(TransactionPayloadInner::V1 {
+                extra_config:
+                    TransactionExtraConfig::V1 {
+                        replay_protection_nonce,
+                        ..
+                    },
+                ..
+            }) => replay_protection_nonce.is_some(),
+            _ => false,
+        }
+    }
+
+    pub fn executable(&self) -> TransactionExecutable {
+        // TODO: This function clones the payload. Check if there is a better way.
+        match self {
+            TransactionPayloadWrapper::EntryFunction(entry_function) => {
+                TransactionExecutable::EntryFunction(entry_function.clone())
+            },
+            TransactionPayloadWrapper::Script(script) => {
+                TransactionExecutable::Script(script.clone())
+            },
+            TransactionPayloadWrapper::Multisig(multisig) => multisig.as_transaction_executable(),
+            TransactionPayloadWrapper::Payload(TransactionPayloadInner::V1 {
+                executable, ..
+            }) => executable.clone(),
+            TransactionPayloadWrapper::ModuleBundle(_) => {
+                unimplemented!("ModuleBundle variant is deprecated")
```

Comment:
I'm not sure if we can panic here, given this is user input

Reply:
This function is currently used in SDK for my tests, and in Aptos VM to extract the executable after the prologue is executed. Prologue will discard the ModuleBundle transactions by then. 
If we want to avoid using implemented!, then I can let this function return Result<TransactionExecutable>.